diff -up urlgrabber-3.9.1/urlgrabber/grabber.py.old urlgrabber-3.9.1/urlgrabber/grabber.py
--- urlgrabber-3.9.1/urlgrabber/grabber.py.old	2015-12-21 14:42:34.599038193 +0100
+++ urlgrabber-3.9.1/urlgrabber/grabber.py	2015-12-21 14:49:50.643402327 +0100
@@ -248,6 +248,13 @@ GENERAL ARGUMENTS (kwargs)
 
     Maximum size (in bytes) of the headers.
     
+  ftp_disable_epsv = False
+
+    False, True
+
+    This options disables Extended Passive Mode (the EPSV command)
+    which does not work correctly on some buggy ftp servers.
+
 
 RETRY RELATED ARGUMENTS
 
@@ -833,6 +840,7 @@ class URLGrabberOptions:
         self.size = None # if we know how big the thing we're getting is going
                          # to be. this is ultimately a MAXIMUM size for the file
         self.max_header_size = 2097152 #2mb seems reasonable for maximum header size
+        self.ftp_disable_epsv = False
         
     def __repr__(self):
         return self.format()
@@ -1252,6 +1260,10 @@ class PyCurlFileObject():
         if opts.data:
             self.curl_obj.setopt(pycurl.POST, True)
             self.curl_obj.setopt(pycurl.POSTFIELDS, self._to_utf8(opts.data))
+
+        # ftp
+        if opts.ftp_disable_epsv:
+            self.curl_obj.setopt(pycurl.FTP_USE_EPSV, False)
             
         # our url
         self.curl_obj.setopt(pycurl.URL, self.url)
